base/utf.jl

```diff
@@ -0,0 +1,994 @@
+const UTF_NO_SURROGATES = 2     # don't accept surrogate pairs in UTF-8/UTF-32
+const UTF_ACCEPT_LONG = 4       # accept long encodings (other than long null in UTF-8)
+
+const UTF_LONG = 1              # Long encodings are present
+const UTF_LATIN1 = 2            # characters in range 0x80-0xFF present
+const UTF_UNICODE2 = 4          # characters in range 0x100-0x7ff present
+const UTF_UNICODE3 = 8          # characters in range 0x800-0xd7ff, 0xe000-0xffff
+const UTF_UNICODE4 = 16         # non-BMP characters present
+const UTF_SURROGATE = 32        # surrogate pairs present
+
+# Get a UTF-8 continuation byte, give error if invalid, and update position and character value
+@inline function get_continuation(ch::UInt32, str, pos)
+    byt::UInt8 = str[pos += 1]
+    !is_valid_continuation(byt) && utf_errfunc(UTF_ERR_CONT, pos, byt)
+    (ch << 6) | (byt & 0x3f), pos
+end
+
+#=
+@doc """
+@brief      Validates and calculates number of characters in a string
+
+@param[in]  str     Vector of UInt8
+@param[in]  options flags to determine error handling (default 0)
+
+@return     (total characters, flags, 4-byte, 3-byte, 2-byte)
+@throws     ArgumentError
+""" ->
+=#
+function check_string_utf8(dat::Vector{UInt8}, options::Integer=0)
+    local byt::UInt8
+    local ch::UInt32, surr::UInt32
+    local totalchar=0, num2byte=0, num3byte=0, num4byte=0, flags::UInt=0
+    pos = 0
+    len = sizeof(dat)
+    @inbounds while pos < len
+        ch = dat[pos += 1]
+        totalchar += 1
+        if ch > 0x7f
+            # Check UTF-8 encoding
+            if ch < 0xe0
+                # 2-byte UTF-8 sequence (i.e. characters 0x80-0x7ff)
+                (pos == len) && utf_errfunc(UTF_ERR_SHORT, pos, ch)
+                ch, pos = get_continuation(ch & 0x3f, dat, pos)
+                if ch > 0x7f
+                    num2byte += 1
+                    flags |= (ch > 0xff) ? UTF_UNICODE2 : UTF_LATIN1
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    flags |= UTF_LONG
+                elseif (ch == 0) && ((options & UTF_NO_LONG_NULL) == 0)
+                    flags |= UTF_LONG
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos, ch)
+                end
+             elseif ch < 0xf0
+                # 3-byte UTF-8 sequence (i.e. characters 0x800-0xffff)
+                (pos + 2 > len) && utf_errfunc(UTF_ERR_SHORT, pos, ch)
+                ch, pos = get_continuation(ch & 0x0f, dat, pos)
+                ch, pos = get_continuation(ch, dat, pos)
+                # check for surrogate pairs, make sure correct
+                if is_surrogate_char(ch)
+                    !is_surrogate_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, pos-2, ch)
+                    # next character *must* be a trailing surrogate character
+                    (pos + 3 > len) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos-2, ch)
+                    byt = dat[pos += 1] ; (byt != 0xed) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, byt)
```

Comment:
no need to declare `byt` local at the top if this is the only place it gets used, would rather put the typeassert here since you've got a similar form `byt::UInt8 = dat[pos += 1]` elsewhere. Your semicolon should also really be a newline here.